test/jdk/java/net/Socket/B8312065.java

```diff
@@ -0,0 +1,249 @@
+        t.join();
+    }
+
+    private static Thread startSendingSignalToThread(long osThreadId) throws InterruptedException {
+        // Setup SIGPROF handler
+        Signal.handle(new Signal("PROF"), (signal) -> {
+            signalReceived = true;
+        });
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        // Send SIGPROF to the thread every second
+        Thread t = new Thread(() -> {
+            while (isSendSignal) {
+                try {
+                    Runtime.getRuntime().exec("kill -SIGPROF " + osThreadId).waitFor();
+                    Thread.sleep(1000);
+                } catch (Exception e) {
+                    e.printStackTrace();
+                } finally {
+                    latch.countDown();
+                }
+            }
+        });
+        t.setDaemon(true);
+        t.start();
+
+        latch.await();
+        if (!signalReceived) {
+            throw new RuntimeException("failed to send signal");
+        }
+        return t;
+    }
+
+    static void test() {
+        Socket socket = null;
+        long startMillis = millisTime();
+        int timeoutMillis = 2000;
+        try {
+            socket = new Socket();
+            connectWithTimeout(socket, timeoutMillis);
+            error("connected successfully!");
+        } catch (SocketTimeoutException socketTimeout) {
+            long duration = millisTime() - startMillis;
+            long min = timeoutMillis - 100;
+            long max = timeoutMillis + 2000;
+            if (duration < min) {
+                error("Duration " + duration + "ms, expected >= " + min + "ms");
+            } else if (duration > max) {
+                error("Duration " + duration + "ms, expected <= " + max + "ms");
+            } else {
+                debug("Passed: Received: " + socketTimeout + ", duration " + duration + " millis");
+                passed();
+            }
+        } catch (Exception exception) {
+            error("Connect timeout test failed", exception);
+        } finally {
+            close(socket);
+        }
+    }
+
+    static void connectWithTimeout(Socket socket, int timeout) throws IOException {
+        // There is no good way to mock SocketTimeoutException, just assume 192.168.255.255 is not in use.
+        socket.connect(new InetSocketAddress("192.168.255.255", 8080), timeout);
+    }
+
+    /**
+     * Returns the current time in milliseconds.
+     */
+    private static long millisTime() {
+        long now = System.nanoTime();
+        return TimeUnit.MILLISECONDS.convert(now, TimeUnit.NANOSECONDS);
+    }
+
+    /**
+     * The JFR event records the OS thread ID, which is a reliable way.
+     * Another way is use /proc/thread-self, but unfortunately, it's not supported until linux 3.17
+     * @return Operating System Thread id
+     * @throws IOException
+     * @throws ParseException
+     */
+    private static long getOSThreadId() throws IOException, ParseException {
+        startJFR();
```

Comment:
You may reuse `libNativeThread` for finding threadId and sending signals. This should make the test a lot simpler.
See `SocketAcceptInterruptTest` and `SocketReadInterruptTest` for example.
JFR is an optional feature, it is not available in Minimal VM.

Reply:
Thanks for your advice. I have updated the test code.